### PR TITLE
Refactor several DateUtils, update (relative) times in trip details and directions form

### LINKED ISSUE #700: "in x minutes" does not update
**Describe the bug**
~When searching for connections and~ In the trip detail view, you can see a "in x minutes"/"now"/"x minutes ago" text. After waiting for e.g. one minute (turning the display off or leaving it on), the displayed text should adapt to the new current time. This does not happen and is especially confusing if you search for a connection, go to the first platform, turn your display off, get on the train (e.g.) and look for your connection after some minutes again, it will still report the time before getting on the train.

**To Reproduce**
1. Search for a connection
2. Select a specific connection within the next minutes ~(optional)~
3. Stay on the same screen for some time (or turn your display off and on again after some time)
4. Notice that the time description does not change over time until reloading.

**Expected behavior**
The string should update according to the current time.

**Versions (please complete the following information):**
 - **Transportr Version:** 2.1.0
 - **Device:** Wileyfox Swift
 - **Android Version:** 9

### PR DESCRIPTION
**Pull request by ialokim**
See individual commits for more information. They might be squashed before merging, but I preferred to keep them separate for easier review.

Fixes #700.

### CODE REVIEW

**Comment by grote** [comment]
Wow, thanks a lot! Do you have screenshots for the UI visible changes?

File: app/src/main/java/de/grobox/transportr/trips/search/DirectionsViewModel.kt [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+    val findGpsLocation = MutableLiveData<FavLocationType?>()
+    val timeUpdate = LiveTrigger()
+    private val _now = MutableLiveData<Boolean>()
+    private val _calendar = MutableLiveData<Calendar?>()
+    private val _products = MutableLiveData<EnumSet<Product>>()
+    private val _isDeparture = MutableLiveData<Boolean>()
+    private val _isExpanded = MutableLiveData<Boolean>()
+    val showTrips = SingleLiveEvent<Void>()
+    @JvmField
+    val topSwipeEnabled = MutableLiveData<Boolean>()
```
Comment (grote) [inline app/src/main/java/de/grobox/transportr/trips/search/DirectionsViewModel.kt:68]: why does this need a JvmField annotation?

File: app/src/main/java/de/grobox/transportr/trips/search/DirectionsViewModel.kt [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+    val queryMoreError: LiveData<String>
+        get() = _tripsRepository.queryMoreError
+
+    init {
+        _now.value = true
+        _calendar.value = Calendar.getInstance()
+        _products.value = EnumSet.allOf(Product::class.java)
+        _isDeparture.value = true
+        _isExpanded.value = false
+        topSwipeEnabled.value = false
```
Comment (grote) [inline app/src/main/java/de/grobox/transportr/trips/search/DirectionsViewModel.kt:199]: You can set these initial values in the live data constructor instead.

File: app/src/main/java/de/grobox/transportr/trips/search/TripsFragment.kt [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+import de.grobox.transportr.ui.LceAnimator
+import de.grobox.transportr.utils.Linkify
+import de.grobox.transportr.utils.TransportrUtils.getDragDistance
+import de.schildbach.pte.dto.Trip
+import kotlinx.android.synthetic.main.fragment_trips.*
+import java.util.regex.Pattern
+import javax.annotation.ParametersAreNonnullByDefault
+import javax.inject.Inject
+
+@ParametersAreNonnullByDefault
```
Comment (grote) [inline app/src/main/java/de/grobox/transportr/trips/search/TripsFragment.kt:49]: I think this annotation isn't needed in Kotlin.

File: app/src/test/java/de/grobox/transportr/utils/DateUtilsTest.kt [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+import org.junit.Assert
+import org.junit.Before
+import org.junit.Test
+import org.mockito.Mock
+import org.mockito.Mockito.*
+import org.mockito.MockitoAnnotations
+import java.util.*
+
+
+class DateUtilsTest {
```
Comment (grote) [inline app/src/test/java/de/grobox/transportr/utils/DateUtilsTest.kt:35]: Nice!

File: app/src/test/java/de/grobox/transportr/utils/DateUtilsTest.kt [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+    @Mock
+    lateinit var resources: Resources
+
+    private fun minuteToMillis(minutes: Float): Long {
+        return (minutes * android.text.format.DateUtils.MINUTE_IN_MILLIS).toLong()
+    }
+
+    @Test
+    fun millisToMinutes() {
+        Assert.assertEquals(0, DateUtils.millisToMinutes(minuteToMillis(0f)))
```
Comment (grote) [inline app/src/test/java/de/grobox/transportr/utils/DateUtilsTest.kt:48]: You could import the methods directly, so removing `Assert.` and `DateUtils.`. Makes the code a bit smaller and easier to read, I'd say.

File: app/src/test/java/de/grobox/transportr/utils/IntentUtilsTest.kt [hunk 6]
```diff
@@ -25,7 +25,7 @@ import org.junit.Assert
 import org.junit.Test
 
 
-class MapViewModelTest {
+class IntentUtilsTest {
```
Comment (grote) [inline app/src/test/java/de/grobox/transportr/utils/IntentUtilsTest.kt:28]: Thanks!

**Review by grote — CHANGES_REQUESTED** [review]
Awesome, thanks for the PR!

File: app/src/main/java/de/grobox/transportr/trips/search/DirectionsViewModel.kt (hunk 1 above)
Comment (ialokim) [reply app/src/main/java/de/grobox/transportr/trips/search/DirectionsViewModel.kt:68]: Good question, I guess the IntelliJ converter added this. Will recheck.

File: app/src/main/java/de/grobox/transportr/trips/search/DirectionsViewModel.kt (hunk 2 above)
Comment (ialokim) [reply app/src/main/java/de/grobox/transportr/trips/search/DirectionsViewModel.kt:199]: True, might be cleaner.

File: app/src/main/java/de/grobox/transportr/trips/search/TripsFragment.kt (hunk 3 above)
Comment (ialokim) [reply app/src/main/java/de/grobox/transportr/trips/search/TripsFragment.kt:49]: Again, going to check. I'm not very familiar with these annotations.

File: app/src/test/java/de/grobox/transportr/utils/DateUtilsTest.kt (hunk 5 above)
Comment (ialokim) [reply app/src/test/java/de/grobox/transportr/utils/DateUtilsTest.kt:48]: Sounds like a good idea :+1:

File: app/src/main/java/de/grobox/transportr/trips/search/DirectionsViewModel.kt (hunk 1 above)
Comment (ialokim) [reply app/src/main/java/de/grobox/transportr/trips/search/DirectionsViewModel.kt:68]: Removed.

File: app/src/main/java/de/grobox/transportr/trips/search/TripsFragment.kt (hunk 3 above)
Comment (ialokim) [reply app/src/main/java/de/grobox/transportr/trips/search/TripsFragment.kt:49]: Removed too.

**Comment by ialokim** [comment]
I've just incorporated your comments.

Here are screenshots before the changes:
old

and after:
new

Note the slightly more natural spacing between "Departure" and "Now" or the date/time. Also note that not everything within a ten minutes time-frame is considered as "Now" anymore, instead it shows relative times up to 30 minutes in past or future.

**Comment by grote** [comment]
> Note the slightly more natural spacing between "Departure" and "Now" or the date/time.

Now that they are so close, I wonder if `Now` should be `now` instead. Or `Departure:` with a colon?

> Also note that not everything within a ten minutes time-frame is considered as "Now" anymore,

Was that the case before? 10 minutes sounds a lot for "now".

> instead it shows relative times up to 30 minutes in past or future.

Do you mean seconds or minutes?

**Comment by ialokim** [comment]
> Now that they are so close, I wonder if `Now` should be `now` instead. Or `Departure:` with a colon?

Good point, I had been thinking about the same thing but forgot it at the end. I would prefer `now` instead of the colon as `in x minutes`/`x minutes ago` is also written in lower-case letters.

> Was that the case before? 10 minutes sounds a lot for "now".

Indeed, see here: https://github.com/grote/Transportr/blob/707bea65f8de38caf29089fdc1508637261c0087/app/src/main/java/de/grobox/transportr/utils/DateUtils.java#L94

> Do you mean seconds or minutes?

I mean minutes. If you search for a trip within the next (or previous) 30 minutes, a relative time information (e.g. `Departure in 25 minutes`) will be shown, before or after that timeframe it falls back to the normal time (e.g. `Departure 10:05 AM`).

**Comment by grote** [comment]
>  I would prefer now instead of the colon

Fine with me



> > Was that the case before? 10 minutes sounds a lot for "now".
>>
> Indeed

😯

> it shows relative times up to 30 minutes in past or future.

I would have shown the relative times for longer, but 30min is also fine.

**Pull request merged**